### DISCUSSION TITLE
Issues with running Highcharts library in NiceGUI

### DISCUSSION THREAD

**Discussion by gengyudong** [zauberzeug/nicegui#936] [Q&A] [answered]
### Question

I understand that one has to pass in a python dict to ui.chart() to generate a highcharts plot. However, I am not sure how to pass the event.click function(shown below) in ui.chart().

```python
Highcharts.chart('container', {
    xAxis: {
        categories: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
    },

    plotOptions: {
        series: {
            cursor: 'pointer',
            point: {
                events: {
                    click: function () {
                        alert('Category: ' + this.category + ', value: ' + this.y);
                    }
                }
            }
        }
    },

    series: [{
        data: [29.9, 71.5, 106.4, 129.2, 144.0, 176.0, 135.6, 148.5, 216.4, 194.1, 95.6, 54.4]
    }]
});
```

**Comment by falkoschindler** [comment, marked as answer]
Unfortunately there is currently no official support for sending JavaScript callbacks to the client. But since you can run arbitrary JavaScript code (once the client is connected) you can use something like the following workaround:

```py
@ui.page('/')
async def page(client: Client):
    chart = ui.chart({
        'xAxis': {
            'categories': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
        },
        'plotOptions': {
            'series': {
                'cursor': 'pointer',
            }
        },
        'series': [{
            'data': [29.9, 71.5, 106.4, 129.2, 144.0, 176.0, 135.6, 148.5, 216.4, 194.1, 95.6, 54.4]
        }]
    })
    await client.connected()
    await ui.run_javascript('''
        const chart = getElement(''' + str(chart.id) + ''').chart;
        chart.update({
            plotOptions: {
                series: {
                    point: {
                        events: {
                            click: function() {
                               alert('Category: ' + this.category + ', value: ' + this.y);
                            }
                        }
                    }
                }
            }
        });
        ''', respond=False)
```

**Reply by clj55** [reply]
I got the error:

> async def page(client: Client):
NameError: name 'Client' is not defined



What do I pass into the `client: Client` parameter?

Edited:
I found the Client (from nicegui import Client)

**Reply by clj55** [reply]
Sometimes I get the 
 `TimeoutError: No connection after 3.0 seconds`
because of `await client.connected()`
How do I resolve this?

**Reply by rodja** [reply]
You can increase the timeout: `await client.connected(timeout=10)`. Are you experiencing this on localhost, or do you have a very slow internet connection? Another cause could be code which blocks the main thread. Make sure all code is non-blocking by using async/await.

**Reply by clj55** [reply]
It works now thank you!
Sometimes I also get the `chart is not defined error` resulting from the `chart.update` even though it's defined in the `const chart = ...`
Or the `this.options.navigator` error 
Really weird that these errors happen on and off

**Comment by aFlocker** [comment]
I ran into the same problem, trying to catch the event of the user zooming in on one chart and having it update a different chart, like seen in this example:
- [Master-detail chart](https://www.highcharts.com/demo/highcharts/dynamic-master-detail)

It wasn't perfect but I was able to catch the master chart's 'mouseup' event and then use ui.run_javascript to check what the new zoom window was:


```python
async def checkZoom():
    minDateEpoch = await ui.run_javascript('window.Highcharts.charts[1].axes[0].min') 
    maxDateEpoch = await ui.run_javascript('window.Highcharts.charts[1].axes[0].max')

    minDate = pd.to_datetime(minDateEpoch, unit='ms').date()
    maxDate = pd.to_datetime(maxDateEpoch, unit='ms').date()
    ...
    ###UPDATE OTHER CHARTS TO ONLY SHOW DATA BETWEEN minDate AND maxDate###
    ...

chart.on('mouseup', checkZoom)
```
it worked pretty well except if the user zoomed by dragging the mouse off the chart element then the mouseup event would never fire...

also I realize now I should have gotten the chart object in javascript by using the getElement( {chart.id} ) way instead of hardcoding the index of the chart
